### PR TITLE
Update rustc-ap-* crates to 407.0.0

### PR DESCRIPTION
**Pull request by topecongiro**
This PR updates `rustc-ap-*` crates to 407.0.0 and resolve breaking changes.

cc @Xanewok and @kngwyu.

### CODE REVIEW

File: tests/source/doc-attrib.rs [hunk 1]
```diff
@@ -47,7 +47,6 @@ pub fn foo() {}
 
 // path attrs
 #[clippy::bar]
-#[clippy::bar=foo]
```
Comment (scampi) [inline tests/source/doc-attrib.rs]: why is this removed ?

File: src/items.rs [hunk 2]
```diff
@@ -2295,158 +2266,47 @@ fn rewrite_args(
...
-        })
-        .collect::<Vec<_>>();
-
-    // Account for sugary self.
-    let mut pre_comment_str = "";
-    let mut post_comment_str = "";
-    let min_args = explicit_self
-        .and_then(|explicit_self| rewrite_explicit_self(explicit_self, args, context))
-        .map_or(1, |self_str| {
-            pre_comment_str = context.snippet(mk_sp(span.lo(), args[0].pat.span.lo()));
```
Comment (scampi) [inline src/items.rs]: I thought this was needed for the fix of issue #3198 but obviously tests pass. Do you know why ?

Comment (topecongiro) [reply src/items.rs]: AFAIK an AST node corresponding to `self` used not to have `Span`, so rustfmt had to craft spans manually. However with current libsyntax the AST node for `self` comes with its `Span`, so we can delegate this work to `write_list`.

File: tests/source/doc-attrib.rs (hunk 1 above)
Comment (topecongiro) [reply tests/source/doc-attrib.rs]: Because now this kind of attribute is considered as an invalid syntax. If we keep this the test just fails during parsing it :stuck_out_tongue_closed_eyes:.

For more details see https://github.com/rust-lang/rust/pull/57367 for the rustc's PR that changed the behavior.

File: src/items.rs (hunk 2 above)
Comment (scampi) [reply src/items.rs]: ok thanks
